x/metadata/keeper/msg_server.go

```diff
@@ -411,3 +411,49 @@ func (k msgServer) DeleteRecordSpecification(
 
 	return &types.MsgDeleteRecordSpecificationResponse{}, nil
 }
+
+func (k msgServer) AddP8EContractSpec(
+	goCtx context.Context,
+	msg *types.MsgAddP8EContractSpecRequest,
+) (*types.MsgAddP8EContractSpecResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	newspec, newrecords, err := types.ConvertP8eContractSpec(&msg.Contractspec, msg.Signers)
```

Comment:
These output specs, records specs are not validated with validatebasic (see other comment about adding this conversion step to the validatebasic flow for this request

Reply:
I added more validations in the convert method and this code is also exercised in the messages validate basic.